File: 49-group-anagrams/group-anagrams.cpp
class Solution {
private:
    string generateKey(string word)
    {
        int freq[26] = {0};
        for(char ch:word)
        {
            freq[ch-'a']++;
        } 
        string key ="";
        for(int i = 0 ; i<26;i++)
        {
            if(freq[i]>0)
            {
                char ch = 'a'+i;
                for(int j = 0 ; j<freq[i];j++)
                {
                    key+=ch;
                }
            }
        }
        return key;
    }
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result ;
        unordered_map<string , vector<string>> mp;

        int n = strs.size();
        for(int i = 0 ; i<n ; i++)
        {
            string key = generateKey(strs[i]);
            mp[key].push_back(strs[i]);
        }

        for(auto &it:mp )
        {
            result.push_back(it.second);
        }
        return result;
    }
};

File: 49-group-anagrams/group-anagrams.java
public class Solution {
    private String generateKey(String word) {
        int[] freq = new int[26];
        for (char ch : word.toCharArray()) {
            freq[ch - 'a']++;
        }
        String key ="";
        for (int i = 0; i < 26; i++) {
            if (freq[i] > 0) {
                char ch = (char) ('a' + i);
                for (int j = 0; j < freq[i]; j++) {
                    key +=ch;
                }
            }
        }
        return key;
    }

    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<>();
        Map<String, List<String>> map = new HashMap<>();

        int n = strs.length;
        for (int i = 0; i < n; i++) {
            String key = generateKey(strs[i]);
            map.computeIfAbsent(key, k -> new ArrayList<>()).add(strs[i]);
        }

        for (List<String> group : map.values()) {
            result.add(group);
        }
        return result;
    }

}